src/style-spec/expression/index.js

```diff
@@ -347,22 +351,23 @@ function findZoomCurve(expression: Expression): Step | Interpolate | ParsingErro
     return result;
 }
 
-import { ColorType, StringType, NumberType, BooleanType, ValueType, array } from './types';
+import { ColorType, StringType, NumberType, BooleanType, ValueType, FormattedType, array } from './types';
 
-function getExpectedType(spec: StylePropertySpecification): Type | null {
+function getExpectedType(spec: StylePropertySpecification): Type {
     const types = {
         color: ColorType,
         string: StringType,
         number: NumberType,
         enum: StringType,
-        boolean: BooleanType
+        boolean: BooleanType,
+        formatted: FormattedType
```

Comment:
I think with this change, every place in the code that checks `text instanceof Formatted` has a dead branch. I can go clean those up in a later PR. There may also be some performance impact here because of the extra bookkeeping the "formatted" path uses, although I'd guess it's pretty minimal.